### ISSUE TITLE
"Game Name is not detected"

### ISSUE THREAD

**Issue by codecat** [discordapp/discord-rpc#161]
I'm trying to implement Rich Presence, but I'm getting the following message from my other computer on my other test account:

<img width="494" alt="screen shot 2018-03-28 at 17 06 58" src="https://user-images.githubusercontent.com/136534/38038158-ae462024-32aa-11e8-9e04-3e354bf3d592.png">

Why does this happen? What are the requirements for the game to be detected? Is it not enough to be already running the game with rich presence turned on?

Both accounts are in the same app:
<img width="216" alt="screen shot 2018-03-28 at 17 09 34" src="https://user-images.githubusercontent.com/136534/38038246-d4de01d4-32aa-11e8-9316-cac17fbd0f0c.png">

**Comment by MinnDevelopment** [comment]
You have to register the executable when initializing. See `Discord_RegisterGame`

**Comment by codecat** [comment]
Ah, I see. I don't think the documentation is super clear on this, as I thought this registration was for starting the game from within Discord?

Will this work properly if passing 1 to `Discord_Initialize`?

**Comment by MinnDevelopment** [comment]
Yes, that causes it to automatically call either `Discord_RegisterSteamGame` if a steamId is set or `Discord_RegisterGame` otherwise.

**Comment by codecat** [comment]
I just set the `autoRegister` param to 1 on client (Mac) & host (Windows), and it's still telling me the same thing.

**Comment by msciotti** [comment]
The registration is for starting the game from within Discord, but it goes hand it hand with "detecting" the game. We don't detect the game is able to be launched from discord, so we say it's not detected.

Passing `1` for `autoRegister` will register the game with Discord the next time `Initialize()` is called, presumably on a game re-start. So you'll have to launch the game at least once on the Mac side, and then the invites should be interactive.

**Comment by codecat** [comment]
Why is that registration done on the second Initialize call?

**Comment by msciotti** [comment]
Sorry, my brain is bad today. I don't mean you have to call initialize twice. I mean that now that you've got `1` for `autoRegister`, the _next time_ your game boots up and calls `Initialize()` on the Mac client, it should be properly registered.

**Comment by codecat** [comment]
With "next time" you mean after I compile it with `1` and run it? :)

**Comment by codecat** [comment]
It's still not working for me, even if I manually call the registration:
```c++
	Discord_Initialize(clientID, &handlers, 1, strSteamAppID.c_str());
	Discord_RegisterSteamGame(clientID, strSteamAppID.c_str());
```
Does this rely on the app to be installed within Steam? I'm doing this on my Mac, and we don't officially have a Mac build yet, so I can't install the game from the Steam client.

**Comment by msciotti** [comment]
Yes, passing the steam id requires the app to be distributed via steam. If you take a look at `src/discord_register_osx.m`:

```
void Discord_RegisterSteamGame(const char* applicationId, const char* steamId)
{
    char command[256];
    snprintf(command, 256, "steam://rungameid/%s", steamId);
    Discord_Register(applicationId, command);
}
```

If you pass a steam id, or call Discord_RegisterSteamGame, we save the "executable" command as `steam://rungameid/<steamId>`. So if that doesn't point to a valid steam game on your platform, it won't work.

A couple questions as we continue:

1. What mac Discord client are you using? Canary/Stable/PTB?

2. What does your json file look like at `~/Library/Application Support/discord/games.json`? Is there anything registered there? Does the application id/client id registered there match the executable/command that should be running?

**Comment by codecat** [comment]
1. I'm running PTB on my Mac.

2. There are 2 folders in Application Support, one being `discord` and one being `discordptb`. In neither of them is a `games.json` file, but in the `discord` folder there is a `games` folder containing a json file with the correct SteamID.

I also tried fooling Steam into installing it (and pretending to be on Windows platform), so running `open steam://rungameid/677120` in my terminal tries to start the game from Steam, so that looks fine.

**Comment by msciotti** [comment]
If you try running on the stable client, does Discord detect the game? I have a hunch we may not be looking in the right places if you aren't on the `discord` client.

**Comment by codecat** [comment]
Just checked, it does not detect the game in the stable build either.

**Comment by msciotti** [comment]
@codecat Were you able to get this resolved, or a repro for the issue? Don't wanna leave you hanging on this.

**Comment by codecat** [comment]
Sorry, haven't had the time to test this yet. Thanks for the reminder, I'll try it tomorrow!

**Comment by codecat** [comment]
Okay, followed all the steps, but unfortunately it did not work.

<img width="1180" alt="screen shot 2018-04-06 at 09 42 44" src="https://user-images.githubusercontent.com/136534/38408715-08300764-397f-11e8-9695-4bfcaaf71b60.png">

**Comment by msciotti** [comment]
This is honestly quite weird. Everything looks like it should be in place. Is this perhaps available in a beta branch on Steam or something that I could download and test locally? From your screenshot, everything looks right, but with that same setup on my machine, the embed is interactive 😕

**Comment by codecat** [comment]
There's currently no public Mac build for Heroes of Hammerwatch yet, nor is there a beta for it yet. However, if you DM me on Discord, Miss#8888 on discord.gg/hammerwatch, I think I can send you some binaries that will work alongside the [`HWR Content`](https://steamdb.info/app/677120/depots/) depot on Steam. (And a Steam key for the game if needed.)

**Comment by StrikerMan780** [comment]
This same bug occurs on windows. Isn't detected for a lot of people, and sometimes they are forced to have the game running before being able to click a "join" button.

**Comment by msciotti** [comment]
@StrikerMan780 gonna try and de-dupe issues as much as possible. At this point, I think it may be more helpful for you to open a separate issue (I think you had one before? Can always re-open or make new). That way we can have a focused discussion on your implementation and avoid sending GitHub notifications to unrelated parties. Thank you!
